lib/mosaic-geomath/src/main/java/org/eclipse/mosaic/lib/spatial/PointCloud.java

```diff
@@ -0,0 +1,317 @@
+     * @return simulation time of {@link PointCloud}
+     */
+    public long getTimestamp() {
+        return timestamp;
+    }
+
+    /**
+     * @return sensor's minimum detectable range in distance units of Phabmacs
+     */
+    public double getMinRange() {
+        return minRange;
+    }
+
+    /**
+     * @return sensor's maximum detectable range in distance units of Phabmacs
+     */
+    public double getMaxRange() {
+        return maxRange;
+    }
+
+    /**
+     * A {@link Point} of the point cloud consists of its coordinates, an identifier
+     * of the type of object the point has hit, and the distance to the point cloud origin.
+     */
+    public static class Point extends Vector3d {
+
+        private static final long serialVersionUID = 1L;
+
+        private final byte hitType;
+        private final float distance;
+
+        /**
+         * @param endPoint the coordinates of the point cloud
+         * @param distance the distance to the origin of the point cloud
+         * @param hitType  the type of hit object represented by this point. 0 = no hit
+         */
+        public Point(Vector3d endPoint, float distance, byte hitType) {
+            x = endPoint.x;
+            y = endPoint.y;
+            z = endPoint.z;
+            this.distance = distance;
+            this.hitType = hitType;
+        }
+
+        /**
+         * @return true if the ray generating this {@link Point} has hit an object
+         */
+        public boolean hasHit() {
+            return hitType != 0;
+        }
+
+        /**
+         * @return the type of the object the ray generating this point has hit. (0 = no hit)
+         */
+        public byte getHitType() {
+            return hitType;
+        }
+
+        /**
+         * @return the distance to the origin of point cloud this point belongs to
+         */
+        public float getDistance() {
+            return distance;
+        }
+
+        @Override
+        public boolean equals(Object o) {
+            if (this == o) {
+                return true;
+            }
+            if (o == null || getClass() != o.getClass()) {
+                return false;
+            }
+            Point other = (Point) o;
+            return super.equals(other) &&
+                    this.hitType == other.hitType &&
+                    Float.compare(this.distance, other.distance) == 0;
+        }
+
+        @Override
+        public int hashCode() {
+            int result = super.hashCode();
+            result = 31 * result + Byte.hashCode(hitType);
+            result = 31 * result + Float.hashCode(distance);
+            return result;
+        }
+    }
+
+    public static class Absolute extends PointCloud {
+
+        private static final long serialVersionUID = 1L;
+
+        private transient List<Point> absoluteEndPointsWithHit = null;
+        private transient List<Point> relativeEndPoints = null;
+        private transient List<Point> relativeEndPointsWithHit = null;
```

Comment:
why do the end points have to be defined in the inheriting classes if they're defined in all inheriting classes anyways?

Reply:
Yes, we only store the points in the abstract `PointCloud` class. Depending on the implementation, these points have a "relative" or "absolute" meaning. If the consumer of a `Relative` point cloud calls `getRelativePoints()`, it just gets the list of stored points, but if she calls `getAbsolutePoints()` they must be converted. To avoid multiple conversions in a row, we just store them transiently in these additional lists (we cache them for re-use).

This whole "relative" vs. "absolute" thing came up through different use-cases in which sometimes it was better to consume only relative end points, sometimes only absolute end points. We additionally could configure the sensor producing the point cloud to either store relative or absolute endpoints. This was to gain maximum performance depending on the use-case but still keep interchangeability.